### DISCUSSION TITLE
Graphql Variable Validation with Jest

### DISCUSSION THREAD

**Discussion by birdwell** [mswjs/msw#501] [Q&A] [answered]
We are wanting to put assertions on variables passed to a graphql query. On the example below, we want to make sure variables are called specifically with certain values.

```
graphql.query<SaveContent, SaveContentVariables>('SaveContent', (req, res, ctx) => {
    return res(
      ctx.data({
        saveContent: {
          success: true,
          id: req.variables.id,
          errors: [],
        },
      })
    );
  })
```

**Comment by kettanaito** [comment, marked as answer]
Hey, @birdwell.

MSW [doesn't encourage assertion on request data](https://mswjs.io/docs/recipes/request-assertions), as it may often be asserted by UI instead. The input data that we sent to the server is often directly or indirectly present in the response that we use to render the UI. If that's possible, strongly consider basing your assertion on the UI that gets rendered by your query response. 

At the same time, we understand there may be cases when the recommendation above is not fully applicable. You can still issue regular assertions within the response resolver:

```js
graphql.query<SaveContent, SaveContentVariables>('SaveContent', (req, res, ctx) => {
  expect(req.variables).toEqual({ id: 'abc-123' })

  return res(
    ctx.data({
      saveContent: {
        success: true,
        id: req.variables.id,
        errors: [],
      },
    })
  );
})
```

Assertion libraries, like jest's `expect`, is basically an `invariant` that will throw upon a falsy match. Any exception in a response resolver will bubble up to your test code and you should see what the exception is about.

**Reply by birdwell** [reply]
@kettanaito Thank you for your reply and thoughts. I'll read through that documentation and see what approach we can take that aligns with MSW and our needs.

**Reply by kettanaito** [reply]
We are always happy to hear from you and see if we can improve on the API regarding your use case. However, it's worth noticing that MSW does not provide any assertion logic, so it's unlikely for us to adopt any kind of API that would directly assert requests/responses. 

We are currently working on a [Life-cycle events API](#482), which may be used to provision request/response based assertions as in your use case.
